lib/src/dbs/node.rs

```diff
@@ -75,4 +75,12 @@ impl Sub<Duration> for Timestamp {
 	}
 }
 
+impl Timestamp {
+	pub(crate) fn get_and_inc(&mut self, rhs: Duration) -> Timestamp {
+		let get = self.clone();
+		self.value += rhs.as_secs();
+		get
```

Comment:
Maybe it's just because I'm a Rust beginner, but; Why do we clone self here even though we take the mutable borrow, and does the increment against `self.value` instead of the cloned `get`?


```suggestion
pub(crate) fn get_and_inc(&mut self, rhs: Duration) -> Timestamp {
		self.value += rhs.as_secs();
		self
}
```

or 

```
pub(crate) fn get_and_inc(&self, rhs: Duration) -> Timestamp {
		let get = self.clone();
		get.value += rhs.as_secs();
		get
}
```

?

Reply:
```
let value = Timestamp(5)
let previous_value = value.get_and_inc(1)
assert(previous_value+1 == value)
```
Basically we want to increase time, while retaining the previous value. Since it is mutable, the previous value must be captured and returned. The new value is a stateful modification. I really don't like this signature, but as I was writing this code that seemed to be the easiest approach at the time. Otherwise we need to capture the result and set it statefully when we might not have access to the fields. It could mean exposing the clock to users (since we need that in the tests).

Reply:
Ahh makes sense! Then can we just reverse it and make it `inc_and_get` instead of `get_and_inc`?
I think the requirement here is to provide a monotonic clock so when to increment isn't actually important?

```
let value = Timestamp(5)
let new_value = value.inc_and_get(1)
assert(value+1 == new_value )
```

Reply:
The difference between `inc and get` and `get and inc` is which value is returned no? Before or post operation. So this is returning the value before operation. So it should be get_and_inc I think

Reply:
@phughk Hey! Sorry for the delayed response! My point was that changing it to `inc and get` (and name it as such) enables us to throw the redundant-in-theory(I believe) `clone` away.

Something like:

```
pub(crate) fn inc_and_get(&self, rhs: Duration) -> &Timestamp {
		self.value += rhs.as_secs();
		self
}
```

But perhaps this won't work actually for whatever reasons I'm not familiar with so... just go ahead!
The original code looks good to me now.
Just reopening this to let you know. Feel free to resolve once confirmed!

Reply:
Removed the function. Rushmore pointed out we can make it Copy. So with the removal of locks, and the removal of cloning, we no longer need this. This was intially introduced for atomicity but tbh, if there are 2 identical timestamps then that isnt a problem. This is test code anyway, so we will find out if the tests break.